### DISCUSSION TITLE
STM32F0 memory usage skyrockets with very little code.

### DISCUSSION THREAD

**Discussion by XanCraft21** [stm32duino/Arduino_Core_STM32#2102] [General]
**Please, before reporting any issue**
- Make sure you are using the latest Arduino_Core_STM32 version.
  https://github.com/stm32duino/Arduino_Core_STM32/releases/latest
- Make sure the issue is not already reported/fixed on GitHub or discussed on the [stm32duino forum](https://stm32duino.com)
- Submit a GitHub issue only for reporting a problem related to the [Arduino_Core_STM32](https://github.com/stm32duino/Arduino_Core_STM32).
- **Avoid** to submit a GitHub issue for project troubleshooting.

Any questions/feedback/suggestions should be discussed on the [stm32duino forum](https://stm32duino.com):
  * questions on the [STM32 Core](https://www.stm32duino.com/viewforum.php?f=35)
  * bugs/enhancements on the [STM core: Bugs and enhancements](https://www.stm32duino.com/viewforum.php?f=38)

&#x26A0; When reporting any issue, please try to provide all relevant information to help on its resolution.

**Describe the bug**
Hello, i am having a problem with memory usage on my STM32F030 board. This is just for compiling and not for uploading. When i write a very small amount of code, i noticed that the used memory after a compile rapidly increases, even just to blink an LED is already at 66% flash usage. And a bare bones sketch already uses 49%. This value increases unusually quickly and i have never seen this happen with my other boards packages before. If anyone can help me figure this out i will gladly appreciate it, thank you foe your help.

**To Reproduce**
There is really no good code to show, just write as you go and frequently verify the code.

1. Start with a bare bones sketch, hit verify or compile, the used flash memory is 49%
2. Write a code to blink the led, 66% is used.
3. Write a code to slowly fade the led with pwm, including a value to keep track of current brightness, a delay value, a speed value, and a boolean value used to invert the fade, 88% used.
4. Or just simply include the Wire library in a bare bones sketch, 79% used.

**Expected behavior**
Memory usage goes up slowly as more code is being added, and not very quickly.

**Screenshots**
Will add soon.

**Desktop (please complete the following information):**
 - OS: Windows
 - Arduino IDE version: 1.8.19
 - STM32 core version: 2.5.0
 - Tools menu settings: default
 - Upload method: Serial (using hacked drivers).

**Board (please complete the following information):**
 - Name: Generic STM32F030F4x
 - Hardware Revision: unsure, the board is a clone from china.
 - Extra hardware used if any: FTDI adapter

**Additional context**
Uploading to the board works fine, it’s just the memory usage on the Arduino side is crazy, even for a very tiny code. I believe my other STM32 boards don’t have this issue. I am not sure if these memory problems are normal, but I believe i’m the only person with this issue. I truly believe this is a real bug, and as far as i can tell no-one else posted anything about this problem. I hope a solution can be made to fix this problem. Thank you again for your help.

**Comment by fpistm** [comment]
See the wiki:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/HAL-configuration#example-of-blinkino-for-nucleo-l031k6-with-32768-bytes-of-flash-and-8192-bytes-of-ram

https://github.com/stm32duino/Arduino_Core_STM32/wiki/Custom-definitions#custom-pinmap-array

**Reply by XanCraft21** [reply]
I did not know i could just disable some of the HAL drivers and get some space back. I’m sorry i wasted an issue.

After some testing i am seeing a big difference. I disabled all of them and noticed i need at least some of them enabled to make certain functions work. Thank you for your help.

Although the board i was using still really can’t be used because even the core drivers suck up the flash usage, so i can’t use the Wire library to drive an oled display.

**Reply by XanCraft21** [reply]
Wait this is not fully resolved yet, i’m still having memory usage issues making the controller useless.

Disabling the HAL drivers makes some features not work. Even if i just use one of the drivers for example the I2C driver, the used memory is still too high to be usable.

I am unsure what to do and i feel that no-one really wants to help me at this time.

**Reply by fpistm** [reply]
Well we cannot do anything. This core is generic for all STM32 which leads to some oeverhead. Anyway, with some customization or using direclty HAL/LL you can drastically reduce size. Redefine peripheralPins arrays, disable all unused HAL, redefine systemCoreClock_config using LL,... Seems you used an external library and it can be also optimized. Check which part of the code uses lot of space using objdump also.

**Reply by fpistm** [reply]
STM32F030F4Px has a very limited flash size: 16k is very small.
Example to get Wire available:

## Config:
* Gerneric STM32F030F4Px
* Serial disabled

## build_opt.h file
`-DHAL_ADC_MODULE_DISABLED -DHAL_RTC_MODULE_DISABLED -DHAL_SPI_MODULE_DISABLED -DHAL_TIM_MODULE_DISABLED -DHAL_EXTI_MODULE_DISABLED
`
## Sketch
```C++
// Wire Master Writer
// by Nicholas Zambetti <http://www.zambetti.com>

// Demonstrates use of the Wire library
// Writes data to an I2C/TWI slave device
// Refer to the "Wire Slave Receiver" example for use with this

// Created 29 March 2006

// This example code is in the public domain.

#include <Wire.h>
#include "stm32yyxx_ll_bus.h"
#include "stm32yyxx_ll_rcc.h"
#include "stm32yyxx_ll_system.h"
#include "stm32yyxx_ll_utils.h"

extern "C" void SystemClock_Config(void)
{
  LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_SYSCFG);
  LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
  LL_FLASH_SetLatency(LL_FLASH_LATENCY_1);
  while(LL_FLASH_GetLatency() != LL_FLASH_LATENCY_1)
  {
  }
  LL_RCC_HSI_Enable();

   /* Wait till HSI is ready */
  while(LL_RCC_HSI_IsReady() != 1)
  {

  }
  LL_RCC_HSI_SetCalibTrimming(16);
  LL_RCC_PLL_ConfigDomain_SYS(LL_RCC_PLLSOURCE_HSI_DIV_2, LL_RCC_PLL_MUL_12);
  LL_RCC_PLL_Enable();

   /* Wait till PLL is ready */
  while(LL_RCC_PLL_IsReady() != 1)
  {

  }
  LL_RCC_SetAHBPrescaler(LL_RCC_SYSCLK_DIV_1);
  LL_RCC_SetAPB1Prescaler(LL_RCC_APB1_DIV_1);
  LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_PLL);

   /* Wait till System clock is ready */
  while(LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_PLL)
  {

  }
  LL_Init1msTick(48000000);
  LL_SetSystemCoreClock(48000000);
}

void setup()
{
  Wire.begin(); // join i2c bus (address optional for master)
}

byte x = 0;

void loop()
{
  Wire.beginTransmission(4); // transmit to device #4
  Wire.write("x is ");        // sends five bytes
  Wire.write(x);              // sends one byte  
  Wire.endTransmission();    // stop transmitting

  x++;
  delay(500);
}
```

## result:

> Sketch uses 11380 bytes (69%) of program storage space. Maximum is 16384 bytes.
> Global variables use 764 bytes (18%) of dynamic memory, leaving 3332 bytes for local variables. Maximum is 4096 bytes.

Serial uses lot of flash so if possible avoid to use it.

**Reply by XanCraft21** [reply]
I did not expect this core to have so much customization. I will do more testing soon. I will see if i can get manually driving a small SSD1306 display to work. So far no luck but i will check out more settings to see what i can do.

Thanks again for your help.

**Reply by XanCraft21** [reply]
I got the I2C working. I was able to remove the large function at the top and some of the library includes and still get the desired result, but i had to increase the I2C clock speed a lot.

although i can’t use a pre-made library to drive the display i am using i was able to do it manually. Thanks again for your help.
